lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp

```diff
@@ -0,0 +1,126 @@
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/LLVM.h"
+#include "circt/Support/Namespace.h"
+#include "circt/Support/SymCache.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/Support/Debug.h"
+
+#include <mutex>
+
+#define DEBUG_TYPE "lower-hw-instance-choices"
+
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+using namespace sv;
+using namespace ExportVerilog;
+
+namespace {
+
+struct HWLowerInstanceChoicesPass
+    : public HWLowerInstanceChoicesBase<HWLowerInstanceChoicesPass> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+LogicalResult
+ExportVerilog::lowerHWInstanceChoices(mlir::ModuleOp module,
+                                      const LoweringOptions &options) {
+  // Collect all instance choices & symbols.
+  SmallVector<InstanceChoiceOp> instances;
+  SymbolCache symCache;
+  {
+    for (Operation &op : *module.getBody()) {
+      if (auto module = dyn_cast<HWModuleOp>(&op)) {
+        module.walk([&](InstanceChoiceOp inst) { instances.push_back(inst); });
+        symCache.addSymbol(module);
+        continue;
+      }
+      if (auto sym = dyn_cast<SymbolOpInterface>(&op)) {
+        symCache.addSymbol(sym);
+        continue;
+      }
+    }
+  }
+
+  // Build a namespace to generate unique macro names.
+  Namespace ns;
+  ns.add(symCache);
+
+  auto declBuilder = OpBuilder::atBlockBegin(module.getBody());
+  for (InstanceChoiceOp inst : instances) {
+    auto parent = inst->getParentOfType<HWModuleOp>();
+
+    auto defaultModuleOp = cast<HWModuleLike>(
+        symCache.getDefinition(inst.getDefaultModuleNameAttr()));
+
+    // Generate a macro name to describe the target of this instance.
+    std::string name;
+    {
+      llvm::raw_string_ostream os(name);
+      os << "__circt_choice_" << parent.getName() << "_"
+         << inst.getInstanceName();
+    }
+
+    auto symName = ns.newName(name);
+    auto symNameAttr = declBuilder.getStringAttr(symName);
+    auto symRef = FlatSymbolRefAttr::get(symNameAttr);
+    declBuilder.create<MacroDeclOp>(inst.getLoc(), symNameAttr,
+                                    /*args=*/ArrayAttr{},
+                                    /*verilogName=*/StringAttr{});
+
+    // This pass now generates the macros and attaches them to the instance
+    // choice as an attribute. As a better solution, this pass should be moved
+    // out of the umbrella of ExportVerilog and it should lower the `hw`
+    // instance choices to a better SV-level representation of the operation.
+    ImplicitLocOpBuilder builder(inst.getLoc(), inst);
+    builder.create<sv::IfDefOp>(
+        symName, [&] {},
+        [&] {
+          builder.create<sv::MacroDefOp>(
+              symRef, builder.getStringAttr("{{0}}"),
+              builder.getArrayAttr(
+                  {FlatSymbolRefAttr::get(defaultModuleOp.getNameAttr())}));
+        });
```

Comment:
Consider that the PR can re-use the builder it already has, moving the insertion point, and resetting it with a guard. I think the following works. `declBuilder` is used for consistency, but this is really just `builder` with the patch below applied:

```suggestion
    OpBuilder::InsertionGuard guard(declBuilder);
    declBuilder.setInsertionPoint(inst);
    declBuilder.create<sv::IfDefOp>(inst.getLoc(),
        symName, [&] {},
        [&] {
          builder.create<sv::MacroDefOp>(
              symRef, builder.getStringAttr("{{0}}"),
              builder.getArrayAttr(
                  {FlatSymbolRefAttr::get(defaultModuleOp.getNameAttr())}));
        });
```